doc/operator-notes.md

```diff
@@ -30,3 +30,34 @@ In the second two cases, where there is a preexisting filesystem, Ignition's beh
+
+## Partition Reuse Semantics
+Similar to the filesystemd use semantics, there are three possibilties for partitions when Ignition runs
+
+- The is no preexisting partition
+- There is a preexisting partition of the correct size, starting sector, label, type_guid and guid.
+- There is a preexisting partition of an incorrect size, starting sector, label, type_guid or guid.
+
+In the first case, Ignition will always create the desired partition. Using a partition number of 0 specifies to use the next available partition number and will ensure there is no preexisting partition.
+
+In the second two cases, the wipePartition flag controls Ignition's behavior. If it is set to true, Ignition will first delete the partition with the specified number and then attempt to create the new partition. The partition number must be specified and non-zero if wipePartition is set to true.
+
+If wipePartition is false, Ignition will attempt to reuse the existing partition. If the partition has a matching label, uuid, type_guid, start sector, and size, Ignition will reuse the partition. If any of those are not specified, zero, or empty, they do not need to match for Ignition to reuse the partition. If the partition does not match the Ignition configuration, Ignition will fail, and the machine will fail to boot.
+
+## Raid Reuse Semantics
+Just like filesystems and partitions, there are three possibilities for raid arrays when Ignition runs
+
+- The devices specified by the RAID array are unused
+- The devices specified by the RAID array are used in an existing RAID array with the correct level, devices, and name
+- The devices specified by the RAID array are used are either not a member of an existing array or are a member of an exisiting RAID array with incorrect level, devices, or name
+
+An unused device is
+ * Not part of an exising RAID array
+ * A partition with no filesystem present
+ * A whole-disk device with not partition table
+
+In the first case, Ignition will always create the raid array.
+
+In the second and third case the overwriteDevices flag controls Ignition's behavior. If it is set to true, Ignition will use the devices specified whether or not they are currently in a RAID array or have an existing filesystem or partition table.
+
+If overwriteDevices is false, Ignition will attempt to use an existing RAID device. The existing RAID device must have the same name and raid level. Additionally it must be composed of exactly the devices specified and all of those devices must share the same array uuid. The number of spares is allowed to differ. If Ignition cannot reuse the existing RAID device, it will fail and the machine will fail to boot.
```

Comment:
To be clear: if I ask for a raid1 of 3 devices with no spares, and point to three devices with the correct name and level and a common UUID, and they claim that the array is composed of 4 devices, will Ignition succeed or fail?

Reply:
Fail. Spares make this more complicated though, especially if we allow `missing` since Ignition runs `mdadm` with `--raid-devices=<(number of devices listed) - (number of spares)>`.

It doesn't appear that `mdadm` reports anything differently if a `missing` device is a spare than if it just wasn't specified at all. I guess it would be only useful in the case of starting an array in a degraded state? We also currently don't support `missing` devices (since it will get stuck in a udev wait loop)

Given that, I say we _do_ count spares as well, just not caring which devices are marked spare. If we add missing, just skip over that device running `mdadm --examine` and subtract 1 from the device count for each `missing` device

In short for each component listed in a `storage.raid` entry that is not `missing`, run `mdadm --examine` and verify:
 * `MD_NAME` (name) must match spec
 * `MD_UUID` (raid array uuid )must be the same as all the other components (and the spec if we decide to add that)
 * `MD_DEVICES` (number of active raid devices) = (number of devices in list that are not `missing`)  - (number of spares specified)
 * `MD_LEVEL` (raid level) matches spec

Reply:
I'm going to take your proposal as a starting point, and step back a bit.

It seems to me there's a basic conflict between the fundamental goals of Ignition (to initialize a system to a known state, or else fail) and those of RAID (to keep the system working after a hardware failure).  In the PXE case, if we require properties of the system which are not strictly necessary (e.g. the correct number of spares) we will cause unproductive boot failures.

There are two cases that seem relevant here:
1. Creating a RAID in the presence of failed disks, and
2. Accepting an existing RAID with failed disks.

The former is possible (e.g. by automatically setting `missing`) but messy, and silently ignoring an unexpected condition seems contrary to the spirit of Ignition.  The latter is necessary, though, if we want to allow running Ignition on every boot.

Handling the latter requires us to be quite aggressive:
1. Since failed drives may not show up on the bus at all, we'd want to ignore those as long as we have a quorum.
2. We shouldn't care which devices are marked spare.  We can check that spares are array members, but that has failure modes: suppose a failed disk is replaced but not administratively re-added to the array.  We would still be booting after the failure, because of the previous point, but *replacing* the disk would cause boot failures.
3. We should accept an existing degraded array after a timeout so long as it's runnable, since that's what udev does (with a 30-second timeout).

We may not want to do all of those things (in the limit, we'd be reimplementing the mdadm assembly logic: how do we know when a degraded array is runnable?) but we should figure out where to draw the line.

By the by, I'm now thinking supporting `missing` in the Ignition config might be a mistake.  It's a useful tool for doing clever RAID rearrangements (e.g. migrating data between RAID arrays in a machine without enough drive slots), but the key point is that it's one in a series of administrative steps toward a goal state, and Ignition notionally runs once.  `missing` is for creating arrays in degraded state, which is, by definition, not a state intended for production.

Reply:
> if we want to allow running Ignition on every boot.

I don't think we do.

> By the by, I'm now thinking supporting missing in the Ignition config might be a mistake.

I'm 100% on board with not supporting `missing`. As you pointed out, it's probably not a good idea to want to do in the first place, and it brings a lot more complexity.

Reply:
>> if we want to allow running Ignition on every boot.

> I don't think we do.

Capturing offline discussion: we do, for the case where a PXE-booted system wants persistent root or data disks.  That's why we needed the `wipeFilesystem` semantics in 2.1.

Reply:
Summarizing offline discussion: we don't ever want to create a RAID if disks are missing, because Ignition's general contract is to fail if anything is wrong.  We do want to accept missing drives in an existing RAID, though.  This leads to a nice decision chain that lets us avoid a lot of logic in Ignition:

1. `overwriteDevices` true?  Require everything to exist.
2. `overwriteDevices` false, all devices present, and all devices unused?  Create RAID.
3. Otherwise, we have no intention to create the RAID; we're only interested in whether it's consistent.  `mdadm` already has logic to verify this, so just start the array and fail if it fails.

We'll need to verify that `mdadm` catches all the error cases we care about, but I expect that it will.

To handle the missing-device case, we'll need to add a timeout when waiting for devices.  udev already starts a systemd timer that starts RAIDs in degraded mode if all components don't show up within 30 seconds.  We'll need to make sure that timer is disabled on Ignition boots, but as long as our timeout is at least 30 seconds, we won't be breaking anything that isn't already broken on non-Ignition boots.

Reply:
In the case where we start an existing RAID, and it does start we still need to check the name and level are correct. Failed/missing drives are OK as long as the array starts, so not every drive needs to have the correct metadata, but the array needs to report the correct name/level.